File: SngJuni/BOJ/boj1991.cpp
#include <iostream>
#include <unordered_map>

using namespace std;

unordered_map<char, pair<char, char>> graph;

void preorder(char root, unordered_map<char, pair<char, char>>& graph) {
    if (root == '.') return;

    cout << root;
    preorder(graph[root].first, graph);
    preorder(graph[root].second, graph);
}

void inorder(char root, unordered_map<char, pair<char, char>>& graph) {
    if (root == '.') return;
    
    inorder(graph[root].first, graph);
    cout << root;
    inorder(graph[root].second, graph);
}

void postorder(char root, unordered_map<char, pair<char, char>>& graph) {
    if (root == '.') return;
    
    postorder(graph[root].first, graph);
    postorder(graph[root].second, graph);
    cout << root;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int N;
    cin >> N;
    char root, left, right;
    for (int i = 0; i < N; i++) {
        cin >> root >> left >> right;
        graph[root] = {left, right};
    }
    
    root = 'A';
    preorder(root, graph);
    cout << '\n';
    inorder(root, graph);
    cout << '\n';
    postorder(root, graph);

    return 0;
}

File: bigsky/BOJ/Python/boj1991.py
# BOJ1991(D2): 트리 순회
import sys

def preorder(n):
    if n != '.':
        print(n, end='')
        preorder(tree[n][0])
        preorder(tree[n][1])

def inorder(n):
    if n != '.':
        inorder(tree[n][0])
        print(n, end='')
        inorder(tree[n][1])

def postorder(n):
    if n != '.':
        postorder(tree[n][0])
        postorder(tree[n][1])
        print(n, end='')

N = int(sys.stdin.readline())
tree = {}

for _ in range(N):
    root, left, right = sys.stdin.readline().split()
    tree[root] = [left, right]

preorder('A')
print()
inorder('A')
print()
postorder('A')